### PR TITLE
fix(speculative): keep DFlash DDP ranks in lockstep on no-valid-anchor skips

### PR DESCRIPTION
**Pull request by khazic**
## Problem

DFlash's trainer raises `NoValidAnchorsError` when every sample in a micro-batch is too
short to form a block. The training loop caught it with a **per-rank** `continue` (skip, no
backward).

Under DDP this is per-rank and data-dependent. The DDP gradient all-reduce is triggered by
`backward()`, so if one rank skips its backward while another runs it, the NCCL collective
mismatches (the others wait on a reduce that never comes -> watchdog **hang**), or, short of
a hang, the accumulation windows / optimizer-step counts **desync across ranks**. The old
code comment acknowledged this ("pre-filter short samples for multi-rank runs to keep
optimizer steps in lockstep") but nothing enforced it. P-EAGLE already handles the analogous
fully-unsupervised batch with a synced backward (`peagle_trainer.py`); DFlash did not.

## Fix

Decide skip-vs-backward in lockstep. Each micro-batch min-reduces a per-rank
`has_valid_anchor` flag via `_all_ranks_have_valid`: if **any** rank has no valid anchors,
**all** ranks skip that micro-batch together, so no rank issues its backward (and gradient
all-reduce) alone and the accumulation windows stay aligned. The reduce is a tiny `int`
collective, safe inside `no_sync` (which only gates the DDP backward all-reduce).
Single-process runs return the local flag unchanged, so the no-DDP path is byte-for-byte the
old skip.

## Test

- `_all_ranks_have_valid` unit tests: single-process pass-through, and the DDP min-reduce
  (mocked collective) for both the "another rank skipped -> all skip" and "all valid -> all
  run" cases.
- Loop test: a no-valid-anchor micro-batch is skipped (counted, no optimizer step) while the
  rest train; the all-valid case runs every step.
- The DDP no-hang behavior itself needs a multi-GPU server run; it cannot be exercised on
  CPU. The unit tests cover the decision logic and the single-process loop.

### CODE REVIEW

File: nemo_automodel/recipes/llm/train_dflash.py [hunk 1]
```diff
@@ -79,6 +79,24 @@ def _all_reduce_mean(value: torch.Tensor) -> torch.Tensor:
...
+    rank skips its backward (and its gradient all-reduce) while another runs its,
+    the collective mismatches (hang) and the accumulation windows desync. Taking
+    the MIN across ranks makes the skip decision unanimous -- every rank skips the
+    micro-batch unless all of them have something to learn from it. The reduce is
+    a tiny independent collective, safe inside ``no_sync`` (which only gates the
+    DDP backward all-reduce). Single-process runs return the local flag unchanged.
+    """
+    if not is_ddp or not (dist.is_available() and dist.is_initialized()):
+        return bool(local_has_valid)
+    flag = torch.tensor([local_has_valid], device=device, dtype=torch.int32)
```
[the remark this answers is not in the document]
Comment (khazic) [reply nemo_automodel/recipes/llm/train_dflash.py:92]: I don't think that's necessary. The hang this PR fixes is a deterministic consequence of one rank issuing its backward while another skips; the MIN-reduce makes the skip unanimous, which is exactly what the unit tests verify at the decision-logic level. A dedicated multi-GPU repro run isn't a blocker here, the logic that prevents the desync is what's under test. Good to merge as is.

**Review by HuiyingLi — DISMISSED** [review]

**Review by HuiyingLi — APPROVED** [review]
